### ISSUE TITLE
GPflow python package name doesn't follow PEP8

### ISSUE THREAD

**Issue by awav** [GPflow/GPflow#452]
Hello everyone,

According to PEP8 - [packages and modules](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) should have short and lowercase names. The irony is that [contribution](https://github.com/GPflow/GPflow/blob/master/contributing.md) markdown file tells that GPflow code should follow PEP8. But it's not too late to change, as GPflow is not present on public pip/anaconda repositories yet and version is below `1.0.0`. 

The transition can be done in three primitive phases:
1. Inform people that current version will be deprecated and replaced with new one on XX/XX/2017. (Warn of deprecation in Readme file or issue on Github).
2. Provide a script which updates the code with GPflow.
```bash
#!/bin/bash
PROJECT=$1
find "${PROJECT}" -type f -exec sed -i -e 's/GPflow\./gpflow./g' {} \;
find "${PROJECT}" -type f -exec sed -i -e 's/^import GPflow/import gpflow/g' {} \;
find "${PROJECT}" -type f -exec sed -i -e 's/^from GPflow/from gpflow/g' {} \;
```
3. Merge PR into the master.

I have created #451, where I renamed only code/module/package related stuff. The documentation, citings and docstrings refer to GPflow name as before.

**Comment by alexggmatthews** [comment]
I personally think the GPflow typesetting is now too entrenched to change this.

**Comment by awav** [comment]
@alexggmatthews, It is fair enough, and it is also a reason why the script for updating must be available for everyone during transition. I think the renaming will be harmless as GPflow is not available through pip and everyone who wants to get a copy either go to github and can see warning message or download specific SHA of GPflow. I would change package name now and make it standard compliant. But, community opinion is more important and I would like to hear what people think of it.

Example: `ternsorflow` transition from `<1.2` version to `1.2`, when some function interfaces were changed and Google provided a script for updating user code.

**Comment by alexggmatthews** [comment]
An example of a heavily used package that doesn't use this convention is IPython. It doesn't seem to have hurt them to badly.

Another example is GPy. 

Changing the naming convention would clash with the published paper on the package:

http://www.jmlr.org/papers/volume18/16-537/16-537.pdf

**Comment by awav** [comment]
http://www.jmlr.org/papers/volume18/16-537/16-537.pdf doesn't have examples of code, links to the `github` project will be the same. The renaming touches only code not documentation. Agree, there are projects with camelcase-like names and of course it doesn't hurt them (maybe) and that's fine. I do believe that the reason behind lowercase naming requirement is to avoid possible problems when project/package moves from case-sensitive to case-insensitive filesystems (like HFS+) and vice versa. Even new macOS APFS filesystem has case-insensitive mode by default.
I see your point and agree that probably it is better to leave it as is. Others are welcome to share their opinion :)

**Comment by jameshensman** [comment]
I have a mild preference for gpflow, since that follows the pep8 standard. 

To be clear, I think the *project* should be called GPflow (github.com/GPflow/...), but the *package* should be gpflow. This might initially sound problematic, but after consideration I think it's quite a clean solution. The paper gets to remain the same, but we're standards compliant with pep8.

**Comment by awav** [comment]
@jameshensman, @alexggmatthews, @alexisboukouvalas, @markvdw, @vals, @PabloLeon 
This message just pull some attention

**Comment by alexisboukouvalas** [comment]
I agree on the name and better do this prior to setting up the pip package.

**Comment by alexggmatthews** [comment]
On the basis of the discussion, I'm leaning toward changing, but I think we should also canvas the opinions of @javdrher and @icouckuy given the work they have done on GPflowOpt  before we do anything.

**Comment by javdrher** [comment]
I spoke to @icouckuy , for us either option is fine though we tend to prefer renaming to gpflow. Its logical we follow the package naming convention of GPflow so this implies renaming to gpflowopt. We are creating an issue in our repo pointing to this discussion to inform users. As we are pre-release and have a small userbase I assume renaming GPflowOpt is not a problem, but if anyone has a serious problem with it they can post it here.

On a sidenote, before removing our pre-release status we were planning on getting rid of pep8 violations in GPflowOpt as much as we can. Perhaps we should expand the discussion to what level of pep8 compliance we want GPflow projects to achieve. #458 now fixes a range of violations but not all: there are mostly variable naming (model.X, Xnew etc) problems left. Often these names correspond to notations frequently used in papers which is in my opinion more valuable than pep8 compliance.

**Comment by alexggmatthews** [comment]
OK @javdrher thanks for the response. As you suggest I've made a discussion issue for the more general questions in #459 .

**Comment by icouckuy** [comment]
I was wondering when and if the python package rename will be committed (or maybe a missed a commit?).

As we are planning to release GPflowOpt in the coming weeks we will rename our package module too if we are certain this has been done by then. If it will take a little longer we will release as is for the first version.

**Comment by alexggmatthews** [comment]
@icouckuy it looks like we will merge #451 since that seems to be the consensus among the developer community. I notice @awav just closed his PR - I am trying to find out why but otherwise it should go through soon. 

More general discussion on pep8 #459  is ongoing and is unlikely to be decided before next week when @jameshensman returns from abroad. I would advise you not to delay your release for this more general discussion.

**Comment by alexggmatthews** [comment]
Because implementing this requires some special treatment the GPflow team have decided #477 will be merged at a specific time:

**29th August at 1pm UK time.** All PRs afterwards will need to use the new name in the code. All PRs before will use the old naming convention. We don't anticipate that changing between the two naming conventions will be too difficult.

**Comment by javdrher** [comment]
Small request: could the version number of GPflow increase with this merge, and the merge commit get tagged as release on github? GPflowOpt will then depend on that release instead of master (which makes sense for compatibility)

**Comment by alexggmatthews** [comment]
Fixed in #477
